### PR TITLE
Fix #1022: Support for Html list tags and indentation in HtmlParser 

### LINKED ISSUE #1022: Support for <ol> and indentation in HtmlParser
Currently, HthmlParser does not support `<ol>` html tag and also nested indentation is also not supported. 
Add support for `<ol>` tags and nested indentation.

Reference for `<ol>` tag support: https://bitbucket.org/Kuitsi/android-textview-html-list/src/master/app/src/main/java/fi/iki/kuitsi/listtest/
Using the above reference you can easily add a tag handler for `<ol>` tag in Html Parser but this does not support indentation.

You will need to work on `utility` module to fix this issue.
For testing this, you can use `HtmlParserTestActivity` along with `faqs` `faq_answer_1` string resource. 
The outcome of this string should be like this:
Screenshot 2020-04-27 at 2 17 38 PM

### PR DESCRIPTION
**Pull request by veena14cs**
## Explanation
This PR fixes #1022. Adds  support for Html `<ol>,<li> and <ul>` tags with indentation.

### For testing
- Change launch Activity to `HtmlParserTestActivity`

Screenshot

Screenshot_20200616-140612

### CODE REVIEW

**Comment by BenHenning** [comment]
@veena14cs why was this recreated?

**Comment by veena14cs** [comment]
> @veena14cs why was this recreated?

@BenHenning Actually, the approach what I used in previous PR was giving indentation to the 1st line and rest of the lines were aligned from 0th margin. Tab wasn't working as per our need.

Screenshot_20200611-152302

So I removed implementation of `\t` and added some margin span. The comments over there I had resolved but again it was invalid according to changes, that's why I recreated this PR.

Screenshot_20200617-114559

**Comment by BenHenning** [comment]
Hey @veena14cs, unless I'm missing something doesn't this still have the same concerns re: licensing from your last PR? It seems like the solution we're using here is essentially a copy of an existing solution made by someone else. Either we need to depend on that code as a library & ensure it's license compatible with Oppia Android, or we need to implement our own solution.

**Comment by veena14cs** [comment]
> Hey @veena14cs, unless I'm missing something doesn't this still have the same concerns re: licensing from your last PR? It seems like the solution we're using here is essentially a copy of an existing solution made by someone else. Either we need to depend on that code as a library & ensure it's license compatible with Oppia Android, or we need to implement our own solution.

@BenHenning Yes, I understand, but I am not using the same implementation as it is in the repo. If we have to use the same code from that repo, then we can do licensing.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+            listParents.peek().openItem(output)
+          } else {
+            listParents.peek().closeItem(output, listParents.size)
+          }
+        } catch (e: EmptyStackException) {
+          Log.d("CustomTagHandler", "No Such Element Found" + listParents.isEmpty())
+        }
+      }
+      else -> {
+        Log.d("CustomTagHandler", "Found an unsupported tag $tag")
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Remove these log statements.

**Review by rt4914 — COMMENTED** [review]
@veena14cs Just one nit change suggested. Also, write a detailed comment where you can explain how each and every function of `CustomTagHandler` is different from its reference and also if they are similar then reason behind similarity. 
This way @BenHenning can understand you concern behind the similarity of `handleTag` function which you mentioned to me. 

Thanks.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 1 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Done.

**Comment by veena14cs** [comment]
> @veena14cs Just one nit change suggested. Also, write a detailed comment where you can explain how each and every function of `CustomTagHandler` is different from its reference and also if they are similar then reason behind similarity.
> This way @BenHenning can understand you concern behind the similarity of `handleTag` function which you mentioned to me.
> 
> Thanks.

Thanks @rt4914. @BenHenning the main implementation of CustomTagHandler is very different from the reference code. In the reference code it is inserting Marker when an opening `<li>` tag is encountered and Marker is removed when a closing </li> tag is encountered to get the location of the tags. But in this we are not using marker. `handleTag()` is similar as we need to check the open `<ol>/<ul>` tags and add it in parent list to get the list items. 

Please let me know if it clears your concern.

File: app/src/main/java/org/oppia/app/testing/HtmlParserTestActivity.kt [hunk 2]
```diff
@@ -37,5 +39,52 @@ class HtmlParserTestActivity : InjectableAppCompatActivity() {
...
+      ).parseOppiaHtml(
+        rawDummyString2,
+        testHtmlContentOrderedListTextView
+      )
+    testHtmlContentOrderedListTextView.text = htmlResult2
+    val bulletSpans = htmlResult2.getSpans<BulletSpanWithRadius>(0, htmlResult2.length, BulletSpanWithRadius::class.java)
+    val bulletSpan0 = bulletSpans[0] as BulletSpanWithRadius
+    val bulletSpan1 = bulletSpans[1] as BulletSpanWithRadius
+    val bulletSpan2 = bulletSpans[2] as BulletSpanWithRadius
+    Log.d("Bullet start  span", " " +htmlResult2.getSpanStart(bulletSpan0).toLong())
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/testing/HtmlParserTestActivity.kt]: Please remove all debug lines.

File: app/src/main/java/org/oppia/app/testing/HtmlParserTestActivity.kt [hunk 3]
```diff
@@ -37,5 +39,52 @@ class HtmlParserTestActivity : InjectableAppCompatActivity() {
         testHtmlContentTextView
       )
     testHtmlContentTextView.text = htmlResult
+    val testHtmlContentOrderedListTextView: TextView =
+      findViewById(R.id.test_html_content_with_ordered_list_text_view)
+    val rawDummyString2 ="""<ul>
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/testing/HtmlParserTestActivity.kt]: Nit: space after ``=``

File: app/src/sharedTest/java/org/oppia/app/parser/HtmlParserTest.kt [hunk 4]
```diff
@@ -137,30 +137,132 @@ class HtmlParserTest {
...
+    assertThat(htmlResult.getSpanEnd(bulletSpan0)).isEqualTo(htmlResult.getSpanStart(bulletSpan1))
+  }
+
+  @Test
+  fun testHtmlContent_getLeadingMargin_isCorrect() {
+    // Given a span with a certain gap width
+    val span = BulletSpanWithRadius(ApplicationProvider.getApplicationContext<Context>(),GAP_WIDTH)
+
+    // Check that the margin is set correctly
+    val expectedMargin = (2 * BULLET_RADIUS +  GAP_WIDTH)
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/parser/HtmlParserTest.kt]: Nit: one space on either side of ``+``

File: utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    bulletRadius = context.resources.getDimensionPixelSize(R.dimen.bullet_radius)
+    wantColor = false
+    color = 0
+  }
+
+  override fun getLeadingMargin(first: Boolean): Int {
+    return 2 * bulletRadius + gapWidth
+  }
+
+  @SuppressLint("NewApi")
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt]: Why is this needed? Is this solution compatible with SDK 16+?

File: utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+class BulletSpanWithRadius(private val context: Context,gapWidth: Int) : LeadingMarginSpan {
+  private val gapWidth: Int = gapWidth
+  private val bulletRadius: Int
+  private val wantColor: Boolean
+  private val color: Int
+
+  init {
+    bulletRadius = context.resources.getDimensionPixelSize(R.dimen.bullet_radius)
+    wantColor = false
+    color = 0
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt]: Do we need this & the above if they never change?

File: utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+import android.text.Layout
+import android.text.Spanned
+import android.text.style.LeadingMarginSpan
+import org.oppia.util.R
+
+/**
+ * Custom Bullet Span implementation (based on [BulletSpan])
+ * Default implementation doesn't allow for radius modification
+ */
+class BulletSpanWithRadius(private val context: Context,gapWidth: Int) : LeadingMarginSpan {
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt]: Space after ``,``

File: utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    bottom: Int,
+    text: CharSequence,
+    start: Int,
+    end: Int,
+    first: Boolean,
+    l: Layout
+  ) {
+    if ((text as Spanned).getSpanStart(this) == start) {
+      val style = p.style
+      var oldcolor = 0
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt]: ``oldColor``

File: utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+          // Bullet is slightly better to avoid aliasing artifacts on mdpi devices.
+          bulletPath!!.addCircle(
+            0.0f,
+            0.0f,
+            1.2f * bulletRadius,
+            Path.Direction.CW
+          )
+        }
+        c.save()
+        c.translate(x + dir * (bulletRadius * 1.2f + 1), (top + bottom) / 2.0f)
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt]: Please add a comment to explain this math.

File: utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        c.drawPath(bulletPath, p)
+        c.restore()
+      } else {
+        c.drawCircle(
+          x + dir * (bulletRadius + 1).toFloat(),
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt]: Ditto here?

File: utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+      }
+      if (wantColor) {
+        p.color = oldcolor
+      }
+      p.style = style
+    }
+  }
+
+  companion object {
+    private var bulletPath: Path? = null
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt]: Prefer not storing static state like this.

File: utility/src/main/java/org/oppia/util/parser/StringUtils.kt [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+ * StringUtils to handle html tag strings.
+ * [Html.TagHandler] is an interface with a single method: handleTag, which is called
+ * when the parser fails to interpret a HTML tag. Since our HTML now has unknown tags
+ * such as <ordered> and </ordered>, <unordered> and </unordered> and  <listitem> and </listitem>
+ * the handleTag method in the CustomTagHandler() class gets called. We’re going to override handleTag
+ * and tailor it to our use case.
+ */
+object StringUtils {
+  /** Custom tag name for creating ordered lists. This replaces <ol>. */
+  const val OL_TAG = "ordered"
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/StringUtils.kt]: Why not include these directly in the tag handler?

Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/StringUtils.kt]: For the constant values, suggest:
- "ordered_list"
- "unordered_list"
- "list_item"

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        }
+      }
+      LI_TAG -> {
+        try {
+          if (opening) {
+            listParents.peek().openItem(output)
+          } else {
+            listParents.peek().closeItem(output, listParents.size)
+          }
+        } catch (e: EmptyStackException) {
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt:63]: What purpose does this rethrow serve?

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+          throw EmptyStackException()
+        }
+      }
+      else -> {
+        // Found an unsupported tag $tag
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Is this else case needed?

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+        if (opening) {
+          listParents.push(OrderedListTag())
+        } else {
+          listParents.pop()
+        }
+      }
+      LI_TAG -> {
+        try {
+          if (opening) {
+            listParents.peek().openItem(output)
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Suggest ``openListItem`` and ``closeListItem``

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+        }
+      }
+      else -> {
+        // Found an unsupported tag $tag
+      }
+    }
+  }
+
+  /** Abstract super class for [UnorderedListTag] and [OrderedListTag]. */
+  private abstract class ListItemTag {
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Could we use a sealed class for this instead of an abstract class & children? All of the classes that are being created are known here.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+    /**
+     * Opens a new list item.
+     *
+     * @param text
+     */
+    open fun openItem(text: Editable) {
+      if (text.length > 0 && text[text.length - 1] != '\n') {
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: A more Kotlin-idiomatic way of doing this might be:

```kotlin
text.lastOrNull()?.let { text.append("\n") }
```

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+        text.append("\n")
+      }
+      val len = text.length
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Suggest just inlining this value.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+      else -> {
+        // Found an unsupported tag $tag
+      }
+    }
+  }
+
+  /** Abstract super class for [UnorderedListTag] and [OrderedListTag]. */
+  private abstract class ListItemTag {
+    /**
+     * Opens a new list item.
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Here & for other methods in ``ListItemTag``, you can probably omit these KDocs if you used a sealed class since it should help self-document a bit better. We may want to add some additional documentation, but that would be to specifically clarify certain aspects of the classes.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+    }
+  }
+
+  /** Class representing the ordered list (`<ol>`) HTML tag. */
+  private inner class OrderedListTag
+  /** Creates a new `<ol>` with start index of 1. */
+  @JvmOverloads constructor(private var nextIdx: Int = 1) : ListItemTag() {
+    override fun openItem(text: Editable) {
+      super.openItem(text)
+      text.append(Integer.toString(nextIdx++)).append(". ")
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Suggest making this part of a separate abstract class that the open function calls rather than overriding it directly (e.g. ``getNextOpenItemPrefix``).

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+      return arrayOf(
+        LeadingMarginSpan.Standard(listItemIndent * (indentation - 1)),
+        BulletSpanWithRadius(context, bulletMargin)
+      )
+    }
+  }
+
+  /** Class representing the ordered list (`<ol>`) HTML tag. */
+  private inner class OrderedListTag
+  /** Creates a new `<ol>` with start index of 1. */
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: You document constructors using the ``@constructor`` tag (see https://kotlinlang.org/docs/reference/kotlin-doc.html#kdoc-syntax) as part of the class KDoc. Please update.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+      text.removeSpan(listTag)
+      if (where != len) {
+        for (replace in replaces) {
+          text.setSpan(replace, where, len, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
+        }
+      }
+    }
+
+    protected abstract fun getReplaces(text: Editable?, indentation: Int): Array<Any>
+    /** Note: This knows that the last returned object from getSpans() will be the most recently added. */
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt:120]: Space above this KDoc per https://developer.android.com/kotlin/style-guide?authuser=3#vertical.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+    private fun getLast(text: Spanned): ListItemTag? {
+      val listTags = text.getSpans(0, text.length, ListItemTag::class.java)
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Can be simplified:

```kotlin
return text.getSpans(/* start= */ 0, text.length, ListItemTag::class.java).lastOrNull()
```

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+      val where = text.getSpanStart(listTag)
+      text.removeSpan(listTag)
+      if (where != len) {
+        for (replace in replaces) {
+          text.setSpan(replace, where, len, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
+        }
+      }
+    }
+
+    protected abstract fun getReplaces(text: Editable?, indentation: Int): Array<Any>
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: This method is worth adding a KDoc for: what should implementations do? What does it return & how should it be used? Also, can we make the return type more specific than ``Any``?

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 25]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    /**
+     * Closes a list item.
+     *
+     * @param text
+     * @param indentation
+     */
+    fun closeItem(text: Editable, indentation: Int) {
+      if (text.length > 0 && text[text.length - 1] != '\n') {
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Ditto for this conditional (can be simplified).

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+        text.append("\n")
+      }
+      val replaces = getReplaces(text, indentation)
+      val len = text.length
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Suggest inlining this again to provide slightly better context than I think the variable name can.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 27]
```diff
@@ -0,0 +1,158 @@
...
+      val listTag = getLast(text)
+      val where = text.getSpanStart(listTag)
+      text.removeSpan(listTag)
+      if (where != len) {
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Please add a comment explaining why this if check is needed (it may not be clear, and isn't actually clear to me at the moment).

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 28]
```diff
@@ -0,0 +1,158 @@
...
+  private inner class UnorderedListTag : ListItemTag() {
+    override fun getReplaces(
+      text: Editable?,
+      indentation: Int
+    ): Array<Any> {
+      var bulletMargin: Int = indent
+      if (indentation > 1) {
+        bulletMargin = indent - bulletSpan.getLeadingMargin(true)
+        if (indentation > 2) {
+          bulletMargin -= (indentation - 2) * listItemIndent
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Why is this correction needed? What happens if indentation is > 3?

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 29]
```diff
@@ -0,0 +1,158 @@
...
+      text.append(Integer.toString(nextIdx++)).append(". ")
+    }
+
+    override fun getReplaces(
+      text: Editable?,
+      indentation: Int
+    ): Array<Any> {
+      var numberMargin: Int = listItemIndent * (indentation - 1)
+      if (indentation > 2) {
+        numberMargin -= (indentation - 2) * listItemIndent
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Ditto here--what about the indentation > 3 case & why is this correction needed?

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 30]
```diff
@@ -0,0 +1,158 @@
...
+    ): Array<Any> {
+      var bulletMargin: Int = indent
+      if (indentation > 1) {
+        bulletMargin = indent - bulletSpan.getLeadingMargin(true)
+        if (indentation > 2) {
+          bulletMargin -= (indentation - 2) * listItemIndent
+        }
+      }
+      return arrayOf(
+        LeadingMarginSpan.Standard(listItemIndent * (indentation - 1)),
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Why do we need an additional leading margin span when BulletSpanWithRadius is already one?

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt [hunk 31]
```diff
@@ -0,0 +1,158 @@
...
+
+  /** Class representing the unordered list (`<ul>`) HTML tag. */
+  private inner class UnorderedListTag : ListItemTag() {
+    override fun getReplaces(
+      text: Editable?,
+      indentation: Int
+    ): Array<Any> {
+      var bulletMargin: Int = indent
+      if (indentation > 1) {
+        bulletMargin = indent - bulletSpan.getLeadingMargin(true)
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Could we instead create a new bullet span here, compute its leading margin as needed, and then return it as part of the array below? We're creating a span anyway, and this would avoid needing the class ``bulletSpan`` variable.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @veena14cs--I took a more thorough pass on the PR.

File: app/src/main/java/org/oppia/app/testing/HtmlParserTestActivity.kt (hunk 2 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/testing/HtmlParserTestActivity.kt]: Done.

File: app/src/main/java/org/oppia/app/testing/HtmlParserTestActivity.kt (hunk 3 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/testing/HtmlParserTestActivity.kt]: Done.

File: app/src/sharedTest/java/org/oppia/app/parser/HtmlParserTest.kt (hunk 4 above)
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/app/parser/HtmlParserTest.kt]: Done.

File: utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt (hunk 6 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt]: No, its not required, if we are not changing it. Hence, I removed it.

File: utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt (hunk 7 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt]: Done.

File: utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt (hunk 8 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt]: Not required. Hence, removed it.

File: utility/src/main/java/org/oppia/util/parser/StringUtils.kt (hunk 12 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/StringUtils.kt]: This was being used in HtmlParser too. So made a common utlil class.

Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/StringUtils.kt]: Done.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 13 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt:63]: If there is no parent tag I need to handle that exception. Actually there was logger. I removed it and added that statement. This line doesn't make any sense so I removed it and added exception logger.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 14 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Not actually. I can remove this.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 15 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Done.

File: utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt (hunk 5 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt]: I removed this annotation. I am not sure if it works with lower SDK's. There is check added in the code for lower device.

Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt]: Radius parameters do not appear until API level 28.  I found other solution to adjust radius of bullets.

File: utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt (hunk 11 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/BulletSpanWithRadius.kt]: Updated.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 16 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Updated to sealed class. Please take thorough look on this.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 17 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: This is changing the condition and adding two new lines.

Screenshot_20200626-171448

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 19 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Done.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 23 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: done.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 26 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Done.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 18 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Done.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 22 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt:120]: Done.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 24 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Added kdoc and replaced Any with LeadingMarginSpan.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 25 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Done.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 27 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: We need to check if there is span for the text. If not add span.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 30 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: Standard BulletSpan() doesn't allow to control bulletRadius, so making use of BulletSpanWithRadius to modify bullet radius.

File: utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt (hunk 29 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt]: @BenHenning This logic is not going so well with deep nesting.

Screenshot_20200630-165406

It is working fine upto 3 level nesting list items.

**Comment by veena14cs** [comment]
@BenHenning and @rt4914 This logic is not working as expected for deep nesting there is reference of [library](https://github.com/daphliu/android-spannable-list-sample) that I have added in the code. I have tried various solutions, but none of them are working that good. This [library](https://github.com/daphliu/android-spannable-list-sample) we can use for deep nesting and works perfectly.

**Comment by rt4914** [comment]
> @BenHenning and @rt4914 This logic is not working as expected for deep nesting there is reference of [library](https://github.com/daphliu/android-spannable-list-sample) that I have added in the code. I have tried various solutions, but none of them are working that good. This [library](https://github.com/daphliu/android-spannable-list-sample) we can use for deep nesting and works perfectly.

@veena14cs I think we will need to discuss this with @BenHenning regarding the licensing. Unassigning myself for now and will discuss this with Ben.

**Comment by BenHenning** [comment]
Reminder to self to update this with the latest develop.

**Comment by Luffy18346** [comment]
@rt4914 PTAL

**Comment by rt4914** [comment]
> @rt4914 PTAL

As per the above comment, Ben needs to update this PR.

@BenHenning Should we keep this PR open?

**Comment by Luffy18346** [comment]
@BenHenning PTAL

**Comment by anandwana001** [comment]
@BenHenning PTAL

**Comment by BenHenning** [comment]
This is pending work I have staged in another PR. Due to it being open for a long time and lack of time to work on this, closing it for now.
